Assets/BossRoom/Scripts/Client/UI/PopupPanel.cs

```diff
@@ -0,0 +1,77 @@
+using System;
+using UnityEngine;
+using UnityEngine.UI;
+using TMPro;
+using Unity.Multiplayer.Samples.BossRoom.Shared.Infrastructure;
+using Unity.Multiplayer.Samples.BossRoom.Shared.Net.UnityServices.Infrastructure;
+using UnityEngine.SceneManagement;
+
+namespace Unity.Multiplayer.Samples.BossRoom.Visual
+{
+    /// <summary>
+    /// responsible for driving all the functionality of the popup panel players see when connecting to the game
+    /// </summary>
+    public class PopupPanel : MonoBehaviour
+    {
+        [SerializeField]
+        TextMeshProUGUI m_TitleText;
+        [SerializeField]
+        TextMeshProUGUI m_MainText;
+
+        static PopupPanel s_Instance;
+
+        void Awake()
+        {
+            s_Instance = this;
+            ResetState();
+        }
+
+        void OnDestroy()
+        {
+            s_Instance = null;
+        }
+
+        public void OnConfirmClick()
+        {
+            ResetState();
+        }
+
+        /// <summary>
+        /// Helper method to help us reset all state for the popup manager.
+        /// </summary>
+        void ResetState()
+        {
+            m_TitleText.text = string.Empty;
+            m_MainText.text = string.Empty;
+            gameObject.SetActive(false);
+        }
+
+        /// <summary>
+        /// Sets the panel to match the given specifications to notify the player.  If display image is set to true, it will display
+        /// </summary>
+        /// <param name="titleText">The title text at the top of the panel</param>
+        /// <param name="mainText"> The text just under the title- the main body of text</param>
+        /// <param name="confirmFunction"> The function to call when the confirm button is pressed.</param>
+        public static void ShowPopupPanel(string titleText, string mainText, Action confirmFunction = null)
+        {
+            if (s_Instance != null)
```

Comment:
One of the classic issues with popup panel is popup stacking, where you have multiple events triggering popups one after another. We might not need this right now, but I'd still have some assert that checks if the popup was already shown, to have some alarm in case this ever happens.

Reply:
Good point. I think it would still be useful now, it would allow us to use the same popup for the UnityServiceErrorMessage that are being published right now but would overwrite some ConnectionStatusMessages if they were also sent to the PopupPanel

Reply:
dunno if we want to complicate this for GDC though. New features == new bugs. If a popup is already open, we could just log the next popup for example. And fix this on develop

Reply:
Makes sense, I'll create a new ticket to add stacking and include UnityServiceErrorMessages in it